File: programmers/Python/P176962_과제_진행하기.py
def solution(plans):
    answer = []
    tmp = []
    idx=0
    for name, start, playtime in plans:
        hours, minutes = map(int, start.split(":"))
        plans[idx][1] = hours * 60 + minutes
        plans[idx][2] = int(playtime)
        idx+=1

    plans = sorted(plans,key = lambda x: x[1])

    for i in range(0, len(plans) - 1):
        current_end_time = plans[i][1] + plans[i][2]
        next_start_time = plans[i+1][1]

        if current_end_time > next_start_time:
            # 작업 보류
            remaining_time = current_end_time - next_start_time
            tmp.append([plans[i][0], remaining_time])
        else:
            # 현재 과제 완료
            answer.append(plans[i][0])
            # 남은 시간 동안 멈춘 과제 처리
            gap = next_start_time - current_end_time
            while tmp and gap > 0:
                name, remaining_time = tmp.pop()
                if remaining_time <= gap: # 남은시간이 gap이하면 작업완료가능
                    gap -= remaining_time
                    answer.append(name)
                else: # gap보다크면 일부만 진행
                    tmp.append([name, remaining_time - gap])
                    break

    # 마지막 과제 처리
    answer.append(plans[-1][0])
    # 남은 멈춘 과제 처리
    while tmp:
        name, remaining_time = tmp.pop()
        answer.append(name)

    return answer
 

자바풀이는 아래와같다

import java.util.*;

class Solution {

File: programmers/Java/P176962_과제_진행하기.java
import java.util.*;

class Solution {
    public String[] solution(String[][] plans) {
        List<Plan> li = new ArrayList<>();
        Deque<PausedTask> tmp = new ArrayDeque<>();
        List<String> ans = new ArrayList<>();
        for (String[] plan : plans) {
            String[] split = plan[1].split(":");
            int start = Integer.parseInt(split[0]) * 60 + Integer.parseInt(split[1]);
            int playTime = Integer.parseInt(plan[2]);
            li.add(new Plan(plan[0], start, playTime));
        }
        
        li.sort((o1,o2) -> o1.start - o2.start); // 오름차순 정렬
        
        for (int i = 0; i < li.size() - 1; i++) {
            String currentName = li.get(i).name;
            int start = li.get(i).start;
            int playTime = li.get(i).playTime;
            int nextStartTime = li.get(i + 1).start;
            int currentEndTime = start + playTime;
            
            if (currentEndTime > nextStartTime) {
                // 현재 과제가 다음 과제 시작 전에 끝나지 않으면 작업 보류
                int remainingTime = currentEndTime - nextStartTime;
                tmp.push(new PausedTask(currentName, remainingTime));
            } else { // 현재 과제가 다음 과제 시작 전 끝날 때
                ans.add(currentName);
                
                int gap = nextStartTime - currentEndTime;
                while (!tmp.isEmpty() && gap > 0) {
                    PausedTask pausedTask = tmp.pop(); // 가장 최근 멈춘 작업
                    if (pausedTask.remainingTime <= gap) { // 남은시간 내 작업완료
                        gap -= pausedTask.remainingTime;
                        ans.add(pausedTask.name);
                    } else { // 남은 시간이 gap보다 크면 일부만 진행
                        pausedTask.remainingTime -= gap;
                        tmp.push(pausedTask);
                        break;
                    }
                }
            }
            
        }
        
        // 마지막 과제처리
        ans.add(li.get(li.size()-1).name);
        
        // 남은 과제들 처리
        while (!tmp.isEmpty()) {
            PausedTask pausedTask = tmp.pop();
            ans.add(pausedTask.name);
        }
        
        return ans.toArray(new String[0]);
    }
    static class Plan {
        String name;
        int start;
        int playTime;
        
        Plan(String name, int start, int playTime) {
            this.name = name;
            this.start = start;
            this.playTime = playTime;
        }
    }
    static class PausedTask {
        String name;
        int remainingTime;
        
        public PausedTask(String name, int remainingTime) {
            this.name = name;
            this.remainingTime = remainingTime;
        }
    }
}
